cfg/builder/builder_walk.cc

```diff
@@ -368,18 +377,33 @@ BasicBlock *CFGBuilder::walk(CFGContext cctx, ast::Expression *what, BasicBlock
 
                 InlinedVector<LocalRef, 2> args;
                 InlinedVector<core::LocOffsets, 2> argLocs;
-                int argIdx = -1;
-                for (auto &exp : s->args) {
-                    argIdx++;
-                    LocalRef temp;
-                    temp = cctx.newTemporary(core::Names::statTemp());
-                    if (argIdx + 1 == s->args.size() && ast::isa_tree<ast::Hash>(exp)) {
-                        current = walkHash(cctx.withTarget(temp), ast::cast_tree<ast::Hash>(exp), current,
-                                           core::Names::buildKeywordArgs());
-                    } else {
-                        current = walk(cctx.withTarget(temp), exp.get(), current);
-                    }
+                auto [posEnd, kwEnd] = s->kwArgsRange();
+                int argIdx = 0;
+                for (; argIdx < posEnd; ++argIdx) {
+                    auto &exp = s->args[argIdx];
+                    LocalRef temp = cctx.newTemporary(core::Names::statTemp());
+                    current = walk(cctx.withTarget(temp), exp.get(), current);
+                    args.emplace_back(temp);
+                    argLocs.emplace_back(exp->loc);
+                }
+
+                for (; argIdx < kwEnd; argIdx += 2) {
+                    auto &key = s->args[argIdx];
+                    auto &val = s->args[argIdx + 1];
+                    LocalRef keyTmp = cctx.newTemporary(core::Names::hashTemp());
+                    LocalRef valTmp = cctx.newTemporary(core::Names::hashTemp());
+                    current = walk(cctx.withTarget(keyTmp), key.get(), current);
+                    current = walk(cctx.withTarget(valTmp), val.get(), current);
+                    args.emplace_back(keyTmp);
+                    args.emplace_back(valTmp);
+                    argLocs.emplace_back(key->loc);
+                    argLocs.emplace_back(val->loc);
+                }
 
+                if (s->hasKwSplat()) {
```

Comment:
This could also be `kwEnd < s->args.size()`, but this seemed like it was more descriptive. It might also be worth changing `kwArgsRange` to also return a boolean that indicates if a splat was present.